### ISSUE TITLE
More flexible AttributesAssert containsOnly feature

### ISSUE THREAD

**Issue by trask** [open-telemetry/opentelemetry-java#3993]
In the instrumentation repo, the groovy-based attribute assertions keep track of the attributes that were asserted against, and fail if there are any additional attributes outside of those expected ones, which is a nice validation against unexpected attributes.

It would be really helpful if the java-based `AttributesAssert` that we are starting to adopt from this repo had some way to do the same.

I'm not sure the best way to expose/implement this functionality. One option is potentially for `containsOnly` to accept flexible value matching (e.g. `hasEntrySatisfying`).

Some background at https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/4872#issue-1077157302

**Comment by gcatanese** [comment]
@trask How about adding a new `public final AttributesAssert contains(Map.Entry<? extends AttributeKey<?>, ?>... entries)` which asserts each Map.Entry is found (ignoring additional attributes)?
If you agree I can give it a try.

**Comment by trask** [comment]
hey @gcatanese!

what I'm looking for is more like the existing:

```
public final AttributesAssert containsOnly(Map.Entry<? extends AttributeKey<?>, ?>... entries)
```

but the current `containsOnly` only accepts constant values, and for some entries I want to be able to pass in a `Consumer<T> valueCondition` (similar to `hasEntrySatisfying`)

**Comment by anuraaga** [comment]
Thanks for the interest @gcatanese - I think this PR requires some design thought and may not be an easy one to jump into. The logger one should be more straightforward and would be massively appreciated!

**Comment by gcatanese** [comment]
Thank you @trask. I am thinking to add
```
public final <T> AttributesAssert hasEntriesSatisfying(Map.Entry<AttributeKey<T>, Consumer<T>>... valueConditions) {
```
which will assert all `Consumer<T>` conditions are satisfied. For example:
```
.hasEntriesSatisfying(
    new AbstractMap.SimpleImmutableEntry<>(AttributeKey.stringKey("bear"),
       value -> assertThat(value).hasSize(3)),
    new AbstractMap.SimpleImmutableEntry<>(AttributeKey.stringKey("bear"),
       value -> assertThat(value).isEqualTo("mya"))
)
```
Is that acceptable? 

PS
I would also create a helper method (in OpenTelemetryAssertions) to create the entry

**Comment by trask** [comment]
@gcatanese would this work if you have attributes of different types (e.g. one `String` and one `Long`)?

**Comment by gcatanese** [comment]
@trask Uhm, it wouldn't, I guess because AttributeKey is different (String and Long). Any suggestion maybe?

**Comment by anuraaga** [comment]
Hi @gcatanese sorry for the lack of feedback. Had to experiment with code to figure out what's possible with Java and eventually landed on #4173 which solves this issue. Thanks for taking a look at this initially though it wasn't such a simple issue unfortunately.

**Comment by gcatanese** [comment]
@anuraaga OK, I see, I had played around but struggled to find a way to have method overloads supporting multiple types. Next time :-). Thanks for the feedback
